Invalid signature when sending TRANSFER transaction

It appears `TRANSFER` transactions still don't work via the Java driver. Although the schema seems to match the Python driver perfectly, for both `CREATE` and `TRANSFER`, only `CREATE` works. Digging into BigchainDB's mainline code, it seems like the failure happens [here, in the transaction commons, during validation](https://github.com/bigchaindb/bigchaindb/blob/master/bigchaindb/common/transaction.py#L1048).

In my experiments with the Python driver, `TRANSFER` transactions worked fine. So there is something weird going on with Java. I try to build my transaction like so:

```java
// prepare input
final FulFill sourceXfer = new FulFill();
sourceXfer.setTransactionId(createTxn.getId());
sourceXfer.setOutputIndex(0);

// build asset reference
final String assetId = createTxn.getId();

    // prepare a xfer transaction, and sign it
final BigchainDbTransactionBuilder.ITransactionAttributes attrs = BigchainDbTransactionBuilder
    .init()
    .addInput(null, sourceXfer, (EdDSAPublicKey) owner.getKeypair().getPublic())  // owner is the source identity
    .addOutput(amount, (EdDSAPublicKey) target.getKeypair().getPublic())  // target is the destination identity
    .addAssets(assetId, String.class)
    .operation(Operations.TRANSFER);

// attach metadata, if we have it
if (asset.getMetadata() != null)
  attrs.addMetaData(asset.getMetadata());

final Transaction txn = attrs.buildAndSignOnly(
    (EdDSAPublicKey) owner.getKeypair().getPublic(),
    (EdDSAPrivateKey) owner.getKeypair().getPrivate());
```

But, here is what I get from Bigchain in my `docker-compose` logs:
```
mongodb_1     | <timestamp> I NETWORK  [listener] connection accepted from x.x.x.x:yy #14 (10 connections now open)
mongodb_1     | <timestamp> I NETWORK  [conn14] received client metadata from x.x.x.x:yy conn14: { driver: { name: "PyMongo", version: "3.7.1" }, os: { type: "Linux", name: "Linux", architecture: "x86_64", version: "4.9.93-linuxkit-aufs" }, platform: "CPython 3.6.6.final.0" }

bigchaindb_1  | <timestamp> | [ERROR] (bigchaindb.web.views.base) HTTP API error: 400 - POST:/api/v1/transactions - Invalid transaction, could not validate (InvalidSignature): Transaction signature is invalid. (bigchaindb_webapi - pid: 31)
```

Most notably, of course, is the last line there, which says:
`Invalid transaction, could not validate (InvalidSignature): Transaction signature is invalid.`

This builds on the original bug filed by @NikitaLos (#2, *Transaction operation issue*).

--- a/README.md
+++ b/README.md
@@ -72,56 +72,78 @@ BigchainDbConfigBuilder
 
 ### Example: Prepare keys, assets and metadata
 ```java
-//    prepare your keys
+// prepare your keys
 net.i2p.crypto.eddsa.KeyPairGenerator edDsaKpg = new net.i2p.crypto.eddsa.KeyPairGenerator();
 KeyPair keyPair = edDsaKpg.generateKeyPair();
 
-//    New asset
+// New asset
 Map<String, String> assetData = new TreeMap<String, String>() {{
     put("city", "Berlin, DE");
     put("temperature", "22");
     put("datetime", new Date().toString());
 }};
 
-//    New metadata
+// New metadata
 MetaData metaData = new MetaData();
 metaData.setMetaData("what", "My first BigchainDB transaction");
 ```
 
-### Example: Create a Transaction
-```java	
-//    Set up your transaction
-Transaction transaction = BigchainDbTransactionBuilder
-	.init()
-	.addAssets(assetData, TreeMap.class)
-	.addMetaData(metaData)
-	.operation(Operations.CREATE)
-	.buildOnly((EdDSAPublicKey) keyPair.getPublic());
-```
+### Example: Create an Asset
 
-### Example: Create and Sign Transaction
+Performing a CREATE transaction in BigchainDB allocates or issues a digital asset.
+
 ```java
-//    Set up your transaction
-Transaction transaction = BigchainDbTransactionBuilder
-	.init()
-	.addAssets(assetData, TreeMap.class)
-	.addMetaData(metaData)
-	.operation(Operations.CREATE)
-	.buildAndSignOnly((EdDSAPublicKey) keyPair.getPublic(), (EdDSAPrivateKey) keyPair.getPrivate());
-
-```
-
-### Example: Create, Sign and Send a Transaction
-```java
-//    Set up your transaction
-Transaction transaction = BigchainDbTransactionBuilder
+// Set up, sign, and send your transaction
+Transaction createTransaction = BigchainDbTransactionBuilder
 	.init()
 	.addAssets(assetData, TreeMap.class)
 	.addMetaData(metaData)
 	.operation(Operations.CREATE)
 	.buildAndSign((EdDSAPublicKey) keyPair.getPublic(), (EdDSAPrivateKey) keyPair.getPrivate())
 	.sendTransaction();
+```
 
+### Example: Transfer an Asset
+
+Performing a TRANSFER transaction in BigchainDB changes an asset's ownership (or, more accurately, authorized signers):
+
+```java
+// Generate a new keypair to TRANSFER the asset to
+KeyPair targetKeypair = edDsaKpg.generateKeyPair();
+
+// Describe the output you are fulfilling on the previous transaction
+final FulFill spendFrom = new FulFill();
+spendFrom.setTransactionId(createTransaction.getId());
+spendFrom.setOutputIndex(0);
+
+// Change the metadata if you want
+MetaData transferMetadata = new MetaData();
+metaData.setMetaData("what2", "My first BigchainDB transaction");
+
+// the asset's ID is equal to the ID of the transaction that created it
+String assetId = createTransaction.getId();
+
+// By default, the 'amount' of a created digital asset == "1". So we spend "1" in our TRANSFER.
+String amount = "1";
+
+// Use the previous transaction's asset and TRANSFER it
+Transaction transferTransaction = BigchainDbTransactionBuilder
+	.init()
+	.addMetaData(metaData)
+
+	// source keypair is used in the input, because the current owner is "spending" the output to transfer it
+	.addInput(null, spendFrom, (EdDSAPublicKey) keyPair.getPublic())
+
+	// after this transaction, the target 'owns' the asset, so, the new output includes the target's public key
+	.addOutput(output, (EdDSAPublicKey) targetKeypair.getPublic())
+
+	// reference the asset by ID when doing a transfer
+	.addAssets(assetId, String.class)
+	.operation(Operations.TRANSFER)
+
+	// the source key signs the transaction to authorize the transfer
+	.buildAndSign((EdDSAPublicKey) keyPair.getPublic(), (EdDSAPrivateKey) keyPair.getPrivate())
+	.sendTransaction();
 ```
 
 ### Example: Setup Config with Websocket Listener
@@ -140,6 +162,31 @@ BigchainDbConfigBuilder
 	.addToken("app_key", "c929b708177dcc8b9d58180082029b8d")
 	.webSocketMonitor(new MyCustomMonitor())
 	.setup();
+```
+
+### More examples
+
+#### Example: Create a Transaction (without signing and without sending)
+```java
+// Set up your transaction but only build it
+Transaction transaction = BigchainDbTransactionBuilder
+	.init()
+	.addAssets(assetData, TreeMap.class)
+	.addMetaData(metaData)
+	.operation(Operations.CREATE)
+	.buildOnly((EdDSAPublicKey) keyPair.getPublic());
+```
+
+#### Example: Create and Sign Transaction (without sending it to the ledger)
+```java
+//    Set up your transaction
+Transaction transaction = BigchainDbTransactionBuilder
+	.init()
+	.addAssets(assetData, TreeMap.class)
+	.addMetaData(metaData)
+	.operation(Operations.CREATE)
+	.buildAndSignOnly((EdDSAPublicKey) keyPair.getPublic(), (EdDSAPrivateKey) keyPair.getPrivate());
+
 ```
 
 <h2>Api Wrappers</h2>

--- a/src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java
+++ b/src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java
@@ -371,12 +371,26 @@ public class BigchainDbTransactionBuilder {
          */
         private void sign(EdDSAPrivateKey privateKey)
                 throws InvalidKeyException, SignatureException, NoSuchAlgorithmException {
-
             String temp = this.transaction.toHashInput();
-            log.debug("TO BE HASHED ---->\n" + temp + "\n<");
             JsonObject transactionJObject = DriverUtils.makeSelfSortingGson(temp);
 
-            byte[] sha3Hash = DriverUtils.getSha3HashRaw(transactionJObject.toString().getBytes());
+            byte[] sha3Hash;
+            if (Operations.TRANSFER.name().equals(this.transaction.getOperation())) {
+                // it's a transfer operation: make sure to update the hash pre-image with
+                // the fulfilling transaction IDs and output indexes
+                StringBuilder preimage = new StringBuilder(transactionJObject.toString());
+                for (Input in : this.transaction.getInputs()) {
+                    if (in.getFulFills() != null) {
+                        FulFill fulfill = in.getFulFills();
+                        String txBlock = fulfill.getTransactionId() + String.valueOf(fulfill.getOutputIndex());
+                        preimage.append(txBlock);
+                    }
+                }
+                sha3Hash = DriverUtils.getSha3HashRaw(preimage.toString().getBytes());
+            } else {
+                // otherwise, just get the message digest
+                sha3Hash = DriverUtils.getSha3HashRaw(transactionJObject.toString().getBytes());
+            }
 
             // signing the transaction
             Signature edDsaSigner = new EdDSAEngine(MessageDigest.getInstance("SHA-512"));

